pyo3-macros-backend/src/pyclass.rs

```diff
@@ -640,6 +771,203 @@ fn impl_enum(
+        cls,
+        &args,
+        methods_type,
+        complex_enum_default_methods(
+            cls,
+            variants.iter().map(|v| (v.ident, v.python_name(&args))),
+        ),
+        default_slots,
+    )
+    .doc(doc);
+
+    // Need to customize the into_py impl so that it returns the variant PyClass
+    let enum_into_py_impl = {
+        let match_arms: Vec<TokenStream> = variants
+            .iter()
+            .map(|variant| {
+                let variant_ident = variant.ident;
+                let variant_cls = gen_complex_enum_variant_class_ident(cls, variant.ident);
+                quote! {
+                    #cls::#variant_ident { .. } => {
+                        let pyclass_init = _pyo3::PyClassInitializer::from(self).add_subclass(#variant_cls);
+                        let variant_value = _pyo3::Py::new(py, pyclass_init).unwrap();
+                        _pyo3::IntoPy::into_py(variant_value, py)
+                    }
+                }
+            })
+            .collect();
+
+        quote! {
+            impl _pyo3::IntoPy<_pyo3::PyObject> for #cls {
+                fn into_py(self, py: _pyo3::Python) -> _pyo3::PyObject {
+                    match self {
+                        #(#match_arms)*
+                    }
+                }
+            }
+        }
+    };
+
+    let pyclass_impls: TokenStream = vec![
+        impl_builder.impl_pyclass(),
+        impl_builder.impl_extractext(),
+        enum_into_py_impl,
+        impl_builder.impl_pyclassimpl()?,
+        impl_builder.impl_freelist(),
+    ]
+    .into_iter()
+    .collect();
+
+    let mut variant_cls_zsts = vec![];
+    let mut variant_cls_pytypeinfos = vec![];
+    let mut variant_cls_pyclass_impls = vec![];
+    let mut variant_cls_impls = vec![];
+    for variant in &variants {
+        let variant_cls = gen_complex_enum_variant_class_ident(cls, variant.ident);
+
+        let variant_cls_zst = quote! {
+            #[doc(hidden)]
+            #[allow(non_camel_case_types)]
+            struct #variant_cls;
+        };
+        variant_cls_zsts.push(variant_cls_zst);
+
+        // TODO(mkovaxx): propagate variant.options
+        let options: PyClassPyO3Options = parse_quote! {
+            extends = #cls, frozen
+        };
+
+        let variant_args = PyClassArgs {
+            class_kind: PyClassKind::Struct,
+            options,
+            deprecations: Deprecations::new(),
+        };
+
+        let variant_cls_pytypeinfo = impl_pytypeinfo(&variant_cls, &variant_args, None);
+        variant_cls_pytypeinfos.push(variant_cls_pytypeinfo);
+
+        let variant_new = complex_enum_variant_new(cls, variant)?;
+
+        let (variant_cls_impl, field_getters) = impl_complex_enum_variant_cls(cls, variant)?;
+        variant_cls_impls.push(variant_cls_impl);
+
+        let pyclass_impl = PyClassImplsBuilder::new(
+            &variant_cls,
+            &variant_args,
+            PyClassMethodsType::Specialization,
```

Comment:
I think this hard-coding is the source of the clippy failure, you'll have to make sure it matches the corresponding input to `build_py_enum`.

Reply:
Thanks for pointing this out preemptively! This was definitely giving me a hard time figuring out what's going on. Even went down the rabbit hole of reading about the `inventory` crate. 😅